Add natdynlink_hook, a hook called when a module is dynamically loaded

lefessan opened the pull request:
This PR adds a hook `natdynlink_hook`, that is called when a native code module is dynamically loaded. It can be used by Memprof and other profiling tools to adapt to such additions.

lefessan committed:
Add caml_natdynlink_hook, a hook called when a module is dynamically loaded

chambart commented:
This looks ok. Why not call the hook before all the effects on the runtime, i.e. before registering the frame tables ?

damiendoligez commented on byterun/caml/roots.h:
```diff
@@ -39,6 +39,12 @@ CAMLextern void caml_do_local_roots(scanning_action f, char * bottom_of_stack,
+   loaded native code module. */
+CAMLextern void (*natdynlink_hook)(void* handle, char* unit);
+#endif
+
```
Why is it in `roots.h`? What does it have to do with roots? Why not `osdeps.h` or `misc.h`? There's no `natdynlink.h` so you should probably create a `hooks.h` and start gathering all the hooks there.

lefessan commented:
@damiendoligez @chambart 
I added `caml/hooks.h` and renamed the hook in `caml_natdynlink_hook`. For the best time to call the hook, it's not clear for me, it doesn't matter for ocp-memprof, so it could be moved before, but my idea was that we would probably prefer to have the runtime properly initialized before calling the hook for a general usage.
